src/mango/src/mango_cursor.erl

```diff
@@ -44,27 +44,286 @@
+            create_cursor(Db, {UsableIndexes, Trace}, Selector, Opts);
         UserSpecifiedIndex ->
-            create_cursor(Db, UserSpecifiedIndex, Selector, Opts)
+            create_cursor(Db, {UserSpecifiedIndex, Trace}, Selector, Opts)
     end.
 
+-spec enhance_candidates(Candidates, [{#idx{}, properties()}]) -> Candidates when
+    Candidates :: #{#idx{} => properties()}.
+enhance_candidates(Entries, Inputs) ->
+    Combiner =
+        fun(Key, Value1, Value2) ->
+            case Key of
+                usable -> Value1 andalso Value2;
+                reason -> lists:append(Value1, Value2);
+                ranking -> Value1 + Value2
+            end
+        end,
+    lists:foldr(
+        fun({Index, Delta}, Map) ->
+            Updater = fun(Value) -> maps:merge_with(Combiner, Value, Delta) end,
+            maps:update_with(Index, Updater, Map)
+        end,
+        Entries,
+        Inputs
+    ).
+
+-type ranking() :: pos_integer().
+-type properties() ::
+    #{
+        usable := boolean(),
+        ranking := ranking(),
+        reason := [reason()]
+    }.
+
+-spec tag_elems(properties(), sets:set(#idx{})) -> [{#idx{}, properties()}].
+tag_elems(Properties, Set) ->
+    sets:fold(fun(Index, Acc) -> [{Index, Properties} | Acc] end, [], Set).
+
+-type reason_description() ::
+    {name, reason()}.
+-type analysis_attribute() ::
+    {usable, boolean()}
+    | {reasons, [reason_description()]}
+    | {ranking, ranking()}
+    | {covering, boolean()}.
+-type analysis() ::
+    {[analysis_attribute()]}.
+-type candidate_index_attribute() ::
+    {index, #idx{}} | {analysis, analysis()}.
+-type candidate_index() ::
+    {[candidate_index_attribute()]}.
+
+-spec extract_candidate_indexes(#cursor{}) -> [candidate_index()].
+extract_candidate_indexes(Cursor) ->
+    #cursor{trace = Trace, index = Winner, fields = Fields} = Cursor,
+    #{
+        all_indexes := AllIndexes,
+        global_indexes := GlobalIndexes,
+        partition_indexes := PartitionIndexes,
+        usable_indexes := UsableIndexes,
+        usability_map := UsabilityMap,
+        filtered_indexes := FilteredIndexes,
+        indexes_of_type := IndexesOfType
+    } = Trace,
+    % specific to view indexes
+    SortedIndexRanges = maps:get(sorted_index_ranges, Trace, []),
+    % simple difference calculations to determine the results in each stage,
+    % without looking at the implementation
+    PartialIndexes = sets:subtract(AllIndexes, GlobalIndexes),
+    OutOfScopeIndexes = sets:subtract(GlobalIndexes, PartitionIndexes),
+    NotUsableIndexes = sets:subtract(PartitionIndexes, UsableIndexes),
+    ExcludedIndexes = sets:subtract(UsableIndexes, FilteredIndexes),
+    UnfavoredIndexes = sets:subtract(FilteredIndexes, IndexesOfType),
+    % determine rankings
+    UnfavoredIndexesRank = max(1, length(SortedIndexRanges)),
+    {_, [PartialIndexesRank, OutOfScopeIndexesRank, NotUsableIndexesRank, ExcludedIndexesRank]} =
+        lists:foldl(
```

Comment:
Wonder if it would be simpler to have a single accumulator initialized as `[UnfavoredIndexesRank]` then and in the fun do:
 ```
  fun(Indexes, [PrevRank | _] = Acc) ->
      Rank = case ...
  [Rank | Acc]
end
```

Up to you, just a suggestion.

Reply:
Yeah, it makes sense.  I will change it.